### ISSUE TITLE
Questions: Code first, C# client and nullability

### ISSUE THREAD

**Issue by erikljung** [protobuf-net/protobuf-net.Grpc#184]
Hi, I'm pretty new to the whole gRPC/protobuf/C# ecosystem so there might be some missconceptions in this question. A fairly long and complex question that may not belong here at all, but it seems to me that you @mgravell is probably one of the most suited persons to answer this -- since my problem seems rooted in different libraries and concepts (which you seem involved with).

## Background
We're building a backend application with gRPC as the public API. We're using NET5 and _protobuf-net.Grpc.AspNetCore (AddCodeFirstGrpc)_ and _Grpc.AspNetCore.Server (MapGrpcService)_ to wire our _[ServiceContract]_ as gRPC services/endpoints. For testing we're not using the _Grpc.Net.Client.GrpcChannel.CreateGrpcService_ (which I guess it's provided by the _protobuf-net.Grpc package_?). Instead we're using the _SchemaGenerator_ from _protobuf-net.Grpc.Reflection_ to generate _.proto_ (proto3) files which then get built by _Grpc.Tools_ into a "real" client which we use to write our tests. This makes it possible for us to get full integration tests of all parts of our code, we're testing the application with a client generated from the .proto files we provide to our consumers -- and this solution works very well most of the time.

## The problem
We're trying to get nullability to work "as good as possible" across the board. We want to define our contracts/messages in a way so that we "know what we get" inside the application, and in the long run this should also be obvious to .proto consumers. With some custom gRPC interceptors we have got this to work as expected in conjuction with the _SchemaGenerator_ without making any changes to the .proto output. BUT we have some issues with the generated C# gRPC client.

Consider the following _[ProtoContract]_ (partly pseudo):
```
public record SomeRequest(bool? Bool, int? Int, string? String);
```

Which gives the .proto output:
```
message SomeRequest {
   bool bool = 1;
   int32 int = 2;
   string string = 3;
}
```

Which give us a generated client message (with non nullable properties) like:
```
new SomeRequest()
    {
        Bool = true,
        Int = 1,
        String = "foo"
    }
```

In order for the backend (gRPC endpoint) to recive null i just pass `new SomeRequest()` and all properties are interpreted as null. Passing the example above will also get interpreted as expected (true, 1, "foo"). But when passing the following, all properties are interpreted as null:

```
new SomeRequest()
    {
        Bool = false,
        Int = 0,
        String = ""
    }
```
This issue is also described in a bit more detail here https://github.com/grpc/grpc-dotnet/issues/1191 but the answer doesn't cover/solve the problem.

## Questions/conclusion
To me the issue seems to be at the clients end and how it's generated. I've done the same tests with a JavaScript client and BloomRPC and everything is working as expected. But I still need to figure out how to get the same behavior with the C# client.

One solution I found that's just working is to add _optional_ to _.proto_ contract like:
```
message SomeRequest {
   optional bool bool = 1;
   optional int32 int = 2;
   optional string string = 3;
}
```
The issue with this is, how do I best accomplish this with the _SchemaGenerator_? I can't seem to find any simple way to hook into the proto generation (which seems to be in _RuntimeTypeModel/MetaType_ right?). Any pointers here?

I also stumbled upon _RuntimeTypeModel.UseImplicitZeroDefaults_ which looks very much like a solution to what I'm observing, but I can't tell if I was able to set it correctly. Still, I have a working Javascript client which makes me think that this is not the issue, any thoughts?

I also found that by using `import "google/protobuf/wrappers.proto";` and eg. ` google.protobuf.Int32Value` the C# client will get generated with the correct nullable properties, but then I end up with the following error when making requests:
```
 "Status(
    StatusCode="Unknown",
    Detail="Exception was thrown by handler. ProtoException:
    Invalid wire-type (String); this usually means you have over-written
    a file without truncating or setting the length; see https://stackoverflow.com//2152978/23354")"
```
I guess this error is thrown while trying to parse the gRPC request. Are these wrappers not supported in my current setup? Or is there something else that needs to be done in order for this to work?

From what I found, if it was possible, it seems like a combination of *optional* and *wrappers* gives me the expected end result -- correctly interpreted values and a C# client with nullable properties. But this is just thoughts from my very limited perspective of this, and maybe there is other solutions out there.

Thanks a lot if someone read all this, all input is highly appreciated. Have a nice day!

**Comment by mgravell** [comment]
Right; what this comes down to is "presence tracking" in the proto3 schema step, which is a new(ish) feature in proto3 (oddly enough it was always there in proto2). What we *should* now generate in this case is - as you say, the same thing but with `optional` before all the fields (in proto2 `optional` means something more complex, but in proto3 `optional` means "with presence tracking enabled"). Because this feature is new(ish), `SchemaGenerator` has not yet been updated to support that scenario. What we should do, therefore, is: *fix that*.

protobuf-net does not currently have explicit consieration of wrappers.proto; not sure exactly what problem you're seeing there, but: that sounds like a bug.

**Comment by erikljung** [comment]
Thanks a lot for the quick response. Just to clarify: `SchemaGenerator` should (in the future) consider nullable members of `[ProtoContract]` as `optional`, eg `public string? Foo { get; set; } would be be `optional string Foo = 1;`? If so, that's probably what I need. Meanwhile, do you have any idea of a workaround, besides reflection and regex?

Regarding the wrappers, seems that this should be supported according to https://docs.microsoft.com/en-us/aspnet/core/grpc/protobuf?view=aspnetcore-5.0. I've tried with `BoolValue`, `Int32Value`, with the same error,
passing a string with `google.protobuf.StringValue` ends up on the server like _"\n\u0003foo"_ which seems weird. Might there be something else in my setup that's missing/needed that should be in place?

**Comment by mgravell** [comment]
> seems that this should be supported according t

The MS docs are talking about using `protoc` with the Google library; protobuf-net is a completely independent re-implementation of the protobuf specification; any promises made by one may or may not apply to the other, in either direction.

No immediate workaround, but the changes to do it probably aren't huge; the problem is finding the time, since ultimately this is my "hobby" stuff. Are you on a particular timescale?

**Comment by erikljung** [comment]
Right, that makes sense I guess. I'm just a bit overwhelmed by all the concepts I'm trying to put togheter :). No particular timescale that I should stress you with, but it's always "as soon as possible" at all times. I'm just happy that you took your time to answer my questions and if you can find the time to fix the _SchemaGenerator_ at some point that would of course be helpful.

**Comment by cbenard** [comment]
> What we should do, therefore, is: _fix that_.

@mgravell has this been implemented?

**Comment by d1820** [comment]
Whats  missing is in 

```csharp
static Type ApplySubstitutes(Type type)
            {
#pragma warning disable CS0618 // Type or member is obsolete
                if (type == typeof(Empty)) return typeof(WellKnownTypes.Empty);
#pragma warning restore CS0618 // Type or member is obsolete
                if (type == typeof(DateTime)) return typeof(WellKnownTypes.Timestamp);
                if (type == typeof(TimeSpan)) return typeof(WellKnownTypes.Duration);
                return type;
            }
```
Its not also checking for conversion to the Nullable wrappers that is part of WellKnownTypes. This should also implement translation based on this: https://learn.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/protobuf-data-types#nullable-types

![image](https://user-images.githubusercontent.com/377631/196699457-3ed59181-1dc1-4414-9bdb-c63aac1ee442.png)

**Comment by lukedukeus** [comment]
Just ran into this myself, The lack of nullable type conversions is enough of a pain point that I'll just write proto files, and have that generated into C#

**Comment by daniatic** [comment]
@mgravell any progress on this issue? If you have some guiding maybe we can assist with the implementation?
